controllers/clustergroupupgrade_controller.go

```diff
@@ -338,10 +338,10 @@ func (r *ClusterGroupUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.
 				}
 
 				if isUpgradeComplete {
-					err = r.precachingCleanup(ctx, clusterGroupUpgrade)
+					err = r.jobAndViewCleanup(ctx, clusterGroupUpgrade)
```

Comment:
The current pattern is to update CGU status only in this cycle and cleanup happens in the next reconcile. If no particular reason why we have to do this as quickly as possible, maybe stick with the current pattern?

Reply:
Always might be strong word. It's only introduced 6 days ago. It's fine if you want to deal with this in another PR. Not sure how we track things like this. Raise a JIRA?

Reply:
One reason why you might want to move now is that the actionsAfterCompletion.DeleteObjects flag will still honored. Otherwise, you would have to add another if check?

Reply:
Sorry, you are right. I read the history wrong.

Reply:
Can we split the discussion into two parts
1. actionsAfterCompletion.DeleteObjects flag
IIUC this flag is intended to delete policies, which are part of the API exposed to user.  However, managedclusterviews and spoke resources for pre-caching and backup are not part of the API. They are auxiliary and transient resources we don't ask user to create for us. IMHO, they must be always deleted, so this flag should not be checked for them.
1. cleanup that happens in the next reconcile for policies and on the current reconcile for the transient objects.
History-wise, both changes were introduced at about the same time, and weren't coordinated, so we need to choose which change is the pattern we want to adhere to, if any. Why postponing the resource deletion to the next reconcile, or why to hurry and delete the transient objects on this reconcile - are there pros and cons we should consider?

Reply:
I completely agree with @vitus133 on the point 1. For, internal and transient object  i.e. stale jobs, their namespaces and their associated manageclusterview cleanup must be done regardless to whether DeleteObjects has been set to true or false as this flag is only intend to be binded with the policy.

For 2nd point, i think this PR is not scoped to do it. We should discuss about it in a team meeting, and if decided, a JIRA issue can be raised to fix it, if required.

Reply:
Thanks guys. Since I am new to this, everything looks new to me :smiley: Sorry, @sabbir-47 

To be honest, I don't have a strong preference on either points. The only thing I am against is inconsistency. The way it is going now, it feels like two code bases growing on their own in one repo. Let's discuss these two in the next standup or groom meeting. My thoughts right now:

1. I am not sure why it was introduced in the first place. @vitus133 to your point though, I would argue the same even for copied policies as well. Leaving them behind would be problematic if the source ones are updated. Maybe it was for special cases in dev environment only? If we decide to keep the flag, I think it should control everything.
2. From what I just learned in troubleshooting scale test, splitting the cleanup logic into the next reconcile call has some advantages. In general, it's desired for controllers/operators to be more responsive, so it can serve multiple CRs "at the same time". Finishing one CR a bit sooner is a lesser concern. Each reconcile call shouldn't take too long and try to do everything in one shot, considering our single threaded design. The price we pay is slightly higher resource usage on ACM by keeping the CRs for a little bit longer.

Reply:
Thanks for the explanation @jc-rh !
I created stories for the both points above:
1. Clarify the use case of the flag with whoever requested it: https://issues.redhat.com/browse/CNF-5012
2. Make the cleanup consistent: https://issues.redhat.com/browse/CNF-5013
Right now I think we should unblock @sabbir-47  from making the recovery functionality operational.
WDYT?

Reply:
Would be happy to get any review on the code i submitted, otherwise looking for 1 or 2 lgtm before i go on a vacation.

Reply:
/lgtm
Stories are there - we'll do it after the holidays :)